build.gradle.kts

```diff
@@ -362,6 +362,7 @@ val projectsWithDisabledFirBootstrap = coreLibProjects + listOf(
 val gradlePluginProjects = listOf(
     ":kotlin-gradle-plugin",
     ":kotlin-gradle-plugin-api",
+    ":kotlin-gradle-plugin-idea",
```

Comment:
I suggest that you add `-idea-api` project where you have only interfaces, without any implementation. That dependency can be used by the Kotlin IDE plugin. Then you are free to evolve the `-idea` project in any way you'd like. 

See you AGP does that for our models - https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:build-system/builder-model/.

Reply:
Good suggestion! I actually considered this approach as well and actively decided to go with a single module (for now). Here is my reasoning, so far:

- sealing of certain entities (like  `KotlinExternalModelContainer`) seemed a safe measure to control serializable backwards compatibility: By knowing all implementations, I can more easily verify that the classes can still be deserialized on the IDE side. I think the power of the `BackwardsCompatibilityDeserializationTest` would shrink if not sealing this type

- future sealing of interfaces (once we can safely use sealed interfaces in Gradle); Same reasoning as above

- OptIn marker requirement: I right now do only verify binary compatibility for entities that are not marked with `InternalKotlinGradlePluginApi`: I can therefore still pretty freely change the shape of the implementations (except breaking backwards compatible deserialization). Even if I would have another "hidden" -impl module, I think requiring everyone to explicitly opt-in, getting a proper message about binary compatibility is still worth it. At this point, I think just having this classes in this module with the marker is just simpler whilst providing the same safety measures.

- Probably premature: I am a little defensive when it comes to adding module dependencies to KGP (that are not bundled into KGP). The reason is, that I have the feeling that initial sync times suffer from loading plugins when the plugin itself depends on many modules (this, however, is just a subjective feeling and might not be true at all)